src/cephadm/tests/test_cephadm.py

```diff
@@ -861,3 +861,53 @@ def test_get_version_node_exporter(self, _call):
         _call.return_value = '', '{}, version 0.16.1'.format(daemon_type.replace('-', '_')), 0
         version = cd.Monitoring.get_version(ctx, 'container_id', daemon_type)
         assert version == '0.16.1'
+
+    @mock.patch('cephadm.os.fchown')
+    @mock.patch('cephadm.get_parm')
+    @mock.patch('cephadm.makedirs')
+    @mock.patch('cephadm.open')
+    @mock.patch('cephadm.make_log_dir')
+    @mock.patch('cephadm.make_data_dir')
```

Comment:
do you know pyfakefs? 

https://github.com/ceph/ceph/pull/39141/commits/4166eca33dcb242f5f60262033737867ec8fde59

Reply:
While it seems its usage reduces the amount of mocks, it unfortunately does not work well with patching `os.fchown`. I'd have expected that `fake_filesystem.set_uid(0)` fixes that issue, but it doesn't. In addition, it seems to break the application of `debugpy` by default:

```
E       FileNotFoundError: [Errno 2] No such file or directory in the fake filesystem: '/home/user/src/ceph/src/cephadm/.tox/py3/lib/python3.8/site-packages/debugpy/_vendored'
```

The issue could be worked around by pausing the patching, which would always need to be done before one want's to use `debugpy`: https://jmcgeheeiv.github.io/pyfakefs/release/usage.html#suspending-patching

It seems there are a couple other issues with it, which I didn't run into, but which we might eventually when using it:

> A list of Python modules that are known to not work correctly with pyfakefs will be collected here:
>
> multiprocessing has several issues (related to points 1 and 3 above). Currently there are no plans to fix this, but this may change in case of sufficient demand.
>
> subprocess has very similar problems and cannot be used with pyfakefs to start a process. subprocess can either be mocked, if the process is not needed for the test, or patching can be paused to start a process if needed, and resumed afterwards (see this issue).
>
> the Pillow image library does not work with pyfakefs at least if writing JPEG files (see this issue)
>
> pandas (the Python data analysis library) uses its own internal file system access written in C. Thus much of pandas will not work with pyfakefs. Having said that, pyfakefs patches pandas so that many of the read_xxx functions, including read_csv and read_excel, as well as some writer functions, do work with the fake file system. If you use only these functions, pyfakefs will work with pandas.

https://jmcgeheeiv.github.io/pyfakefs/release/usage.html#modules-not-working-with-pyfakefs

All in all, I'd rather not use it and keep debugging for me and possibly others simple and working out of the box.

Reply:
@sebastian-philipp But if it ends up in the source anyway, I can adapt the code to include it.